vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+#include <algorithm>
+#include <cmath>
+#include <cstdio>
+#include <float.h>
+#include <time.h>
+#include <sstream>
+#include <ctime>
+
+#include "reductions.h"
+#include "rand48.h"
+#include "vw.h"
+
+
+using namespace std;
+using namespace LEARNER;
+using namespace VW::config;
+
+namespace memory_tree_ns
+{
+    ///////////////////////Helper//////////////////////////////
+    //////////////////////////////////////////////////////////
+    template<typename T> 
+    void pop_at_index(v_array<T>& array, uint32_t index)
+    {
+        if (index >= array.size()){
+            cout<<"ERROR: index is larger than the size"<<endl;
+            return;
+        }
+        if (index == array.size() - 1){
+            array.pop();
+            return;
+        }
+        for (size_t i = index+1; i < array.size(); i++){
+            array[i-1] = array[i];
+        }
+        array.pop();
+        return;
+    }
+
+    void copy_example_data(example* dst, example* src, int oas = false, bool no_feat = false)
+    { 
+        if (oas == false){
+            dst->l = src->l;
+            dst->l.multi.label = src->l.multi.label;
+        }
+        else{
+            dst->l.multilabels.label_v.delete_v();
+            copy_array(dst->l.multilabels.label_v, src->l.multilabels.label_v);
+	    //cout<<dst->l.multilabels.label_v.size()<<endl;
+        }
+
+        copy_array(dst->tag, src->tag);
+        dst->example_counter = src->example_counter;
+        dst->ft_offset = src->ft_offset;
+
+        if (no_feat == false){
+            copy_array(dst->indices, src->indices);
+            for (namespace_index c : src->indices)
+                dst->feature_space[c].deep_copy_from(src->feature_space[c]);
+            dst->ft_offset = src->ft_offset;
+
+            dst->num_features = src->num_features;
+        }
+        dst->partial_prediction = src->partial_prediction;
+        if (src->passthrough == nullptr) dst->passthrough = nullptr;
+        else
+        { 
+            dst->passthrough = new features;
+            dst->passthrough->deep_copy_from(*src->passthrough);
+        }
+        dst->loss = src->loss;
+        dst->weight = src->weight;
+        dst->total_sum_feat_sq = src->total_sum_feat_sq;
+        dst->confidence = src->confidence;
+        dst->test_only = src->test_only;
+        dst->end_pass = src->end_pass;
+        dst->sorted = src->sorted;
+        dst->in_use = src->in_use;
+    }
+
+    inline void free_example(example* ec)
+    {
+        ec->tag.delete_v();
+        if (ec->passthrough){
+            ec->passthrough->delete_v();
+        }
+        for (int i = 0; i < 256; i++)
+            ec->feature_space[i].delete_v();
+        ec->indices.delete_v();
+        free(ec);
+    }
+
+
+
+    void remove_repeat_features_in_f(features& f, float& total_sum_feat_sq )
+    {
+        for (size_t i = 0; i < f.indicies.size(); i++){
```

Comment:
can you point me to the sort_features function?

Reply:
--sort_features is a command-line input. You can explicitly add it to the pipeline by doing something akin to https://github.com/VowpalWabbit/vowpal_wabbit/blob/70ca7d7a0c21c0115a8d920afa73fe2741b387e1/vowpalwabbit/cb_explore.cc#L300 in your setup method. Then when setup_example is called (e.g. via parser), it will run the code here: https://github.com/VowpalWabbit/vowpal_wabbit/blob/6c065483194c36f0337f184b267486bb865938be/vowpalwabbit/parser.cc#L690

Reply:
ok, I will simply remove this function "remove_repeat_features_in" in the code. It only affects *one* dataset (wikipara). For that specific dataset, I will just in default turn on sort_features in the script.